torch/_lobpcg.py

```diff
@@ -0,0 +1,719 @@
+          d`. Also, creating the diagonal matrix `D` is avoided.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        if d1.is_cuda:
+            # Avoid `test_torch.TestTorchDeviceTypeCUDA.test_lobpcg_basic_cuda_float64 leaked 4096 bytes CUDA memory on device 0`
+            d = torch.zeros((d1.shape[0], 1), dtype=d1.dtype, device=d1.device)
+            d[:, 0] = d1
+        else:
+            d = d1.reshape(d1.shape[0], 1)
+
+        dd = mm(d, _utils.transpose(d))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d
+
+    def _get_svqb(self,
+                  U,     # Tensor
+                  drop,  # bool
+                  tau    # float
+                  ):
+        """Return B-orthonormal U.
+
+        .. note:: When `drop` is `False` then `svqb` is based on the
+                  Algorithm 4 from [DuerschPhD2015] that is a slight
+                  modification of the corresponding algorithm
+                  introduced in [StathopolousWu2002].
+
+        Arguments:
+
+          U (Tensor) : initial approximation, size is (m, n)
+          drop (bool) : when True, drop columns that
+                     contribution to the `span([U])` is small.
+          tau (float) : positive tolerance
+
+        Returns:
+
+          U (Tensor) : B-orthonormal columns (:math:`U^T B U = I`), size
+                       is (m, n1), where `n1 = n` if `drop` is `False,
+                       otherwise `n1 <= n`.
+
+        """
+        # type: (Tensor, bool, float) -> Tensor
+        if torch.numel(U) == 0:
+            return U
+        UBU = _utils.qform(self.B, U)
+        d = UBU.diagonal(0, -2, -1)
+        # detect and drop zero columns from U
+        nz = torch.where(abs(d) != 0.0)
+        assert len(nz) == 1, nz
+        if len(nz[0]) < len(d):
+            U = U[:, nz[0]]
+            if torch.numel(U) == 0:
+                return U
+            UBU = _utils.qform(self.B, U)
+            d = UBU.diagonal(0, -2, -1)
+            nz = torch.where(abs(d) != 0.0)
+            assert len(nz[0]) == len(d)
+
+        D = torch.diag_embed(d ** -0.5)
+        DUBUD = _utils.qform(UBU, D)
+        E, Z = _utils.symeig(DUBUD, eigenvectors=True)
+        t = tau * abs(E).max()
+        if drop:
+            keep = torch.where(E > t)
+            assert len(keep) == 1, keep
+            E = E[keep[0]]
+            Z = Z[:, keep[0]]
+        else:
+            E[(torch.where(E < t))[0]] = t
+        return torch.chain_matmul(U, D, Z, torch.diag_embed(E ** -0.5))
+
```

Comment:
The right-most matrix multiplication with diag(E) can be cast to a broadcasted element-wise product.

Reply:
Fixed.